diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6fa0ce17..c5bdd0ee 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,11 +149,11 @@ if(ENABLE_OPENSSL AND OPENSSL_FOUND)
   set(CMAKE_REQUIRED_INCLUDES   "${OPENSSL_INCLUDE_DIR}")
   set(CMAKE_REQUIRED_LIBRARIES  "${OPENSSL_LIBRARIES}")
   if(WIN32)
-    set(CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}" "ws2_32" "bcrypt")
+    set(CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}" "ws2_32" "bcrypt" "crypt32")
   endif()
   check_symbol_exists(SSL_provide_quic_data "openssl/ssl.h" HAVE_SSL_PROVIDE_QUIC_DATA)
   if(NOT HAVE_SSL_PROVIDE_QUIC_DATA)
-    message(WARNING "Disabling OpenSSL due to lack of QUIC support in ${OPENSSL_LIBRARIES}")
+    message(FATAL_ERROR "Disabling OpenSSL due to lack of QUIC support in ${OPENSSL_LIBRARIES}")
   endif()
   cmake_pop_check_state()
 endif()
